lang/funcs/collection.go

```diff
@@ -497,7 +497,21 @@ var LookupFunc = function.New(&function.Spec{
 			return cty.NilType, fmt.Errorf("lookup() takes two or three arguments, got %d", len(args))
 		}
 
-		return args[0].Type().ElementType(), nil
+		ty := args[0].Type()
+		key := args[1].AsString()
+		switch {
+		case ty.IsObjectType():
+			if ty.HasAttribute(key) {
+				return args[0].GetAttr(key).Type(), nil
+			} else if len(args) == 3 {
+				// if the key isn't found but a default is provided,
+				// return the default type
+				return args[2].Type(), nil
+			}
+			return cty.DynamicPseudoType, nil
```

Comment:
This isn't _required_ because the `Impl` function will take care of it eventually anyway, but if we return an error here then we're more likely to flag this problem earlier (during validation/plan) rather than later (during apply) when things are unknown.

```suggestion
			return cty.DynamicPseudoType, fmt.Errorf("the given object has no attribute %q", key)
```